node/src/reactor/main_reactor.rs

```diff
@@ -960,6 +960,74 @@ impl reactor::Reactor for MainReactor {
+                // successor for the comparison.
+                let era_to_check = era_id.successor();
+                // We read the validators for era_id+1 from the parent of the immediate switch
+                // block.
+                let validators_in_parent = match parent_era_validators.get(&era_to_check) {
+                    Some(validators) => validators,
+                    None => {
+                        return fatal!(
+                            effect_builder,
+                            "couldn't find validators for era {} in parent_era_validators",
+                            era_to_check
+                        )
+                        .ignore();
+                    }
+                };
+                // We also read the validators from the immediate switch block itself.
+                let validators_in_block = match block_era_validators.get(&era_to_check) {
+                    Some(validators) => validators,
+                    None => {
+                        return fatal!(
+                            effect_builder,
+                            "couldn't find validators for era {} in block_era_validators",
+                            era_to_check
+                        )
+                        .ignore();
+                    }
+                };
+                // Decide which validators to use for `era_id` in the validators matrix.
+                let validators_to_register = if validators_in_parent == validators_in_block {
+                    // Nothing interesting happened - register the regular validators, ie. the
+                    // ones stored for `era_id` in the parent of the immediate switch block.
+                    match parent_era_validators.get(&era_id) {
+                        Some(validators) => validators,
+                        None => {
+                            return fatal!(
+                                effect_builder,
+                                "couldn't find validators for era {} in parent_era_validators",
+                                era_id
+                            )
+                            .ignore();
+                        }
+                    }
+                } else {
+                    // We had an upgrade changing the validators! We use the same validators that
+                    // will be used for the era after the immediate switch block, as we can't trust
+                    // the ones we would use normally.
+                    validators_in_block
+                };
+                self.validator_matrix
+                    .register_validator_weights(era_id, validators_to_register.clone());
+                // Crank the reactor so that any synchronizing tasks blocked by the lack of
+                // validators for `era_id` can resume.
+                effect_builder
+                    .immediately()
+                    .event(|_| MainEvent::ReactorCrank)
```

Comment:
Before cranking, we probably should call `handle_validators` on at least the synchronizer, possibly the accumulator, to make sure they have the most up-to-date information.

Reply:
The accumulator is for the blocks that are gossiped, isn't it? In such a case it wouldn't be of much use to call `handle_validators` on it, as this will only ever affect blocks in the past, which wouldn't be gossiped anymore.

Calling it on the synchronizer isn't necessary, because the updated validators get read later, anyway, when the synchronizer retries to sync the block and repeats the check if the validators matrix has the relevant era. It could make sense as an optimization, but I'm not sure if it would even make any difference, as cranking the reactor should call `keep_up_instruction` and that will push the sync process further even without us calling `handle_validators`.

Reply:
In the context we're in now and how we process this event, it makes sense that the synchronizer would be able to do without this (as it is proven to do in the tests which pass on this PR), but I'm making the point we should make behaviors such as updating the validator matrix trigger predictable, which means notifying every interested party about this. I would take it a step further and have a higher-level function for updating the validator matrix which does that plus any necessary event dispatches to ensure that, but it's not necessary.

To the point, I agree it works as is and I'm fine with it merging in this form. Surely not a blocker for me.